[webkitcorepy] Create generic docker utilities

JonWBedard opened the pull request:
#### 9677cac049358081971a5113e5c64de74ab7fbe8
<pre>
[webkitcorepy] Create generic docker utilities
<a href="https://bugs.webkit.org/show_bug.cgi?id=257815">https://bugs.webkit.org/show_bug.cgi?id=257815</a>
rdar://110408458

Reviewed by NOBODY (OOPS!).

Create a generic set of Docker utilities for our various Python services.

* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py:
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/docker.py: Added.
(Docker.Container): Object representing a specific docker container managed from Python.
(Docker.Volume): Object representing a specific docker volume managed from Python.
(Docker.command): Return path to Docker command, if it exists.
(Docker.installed): Check if Docker is installed.
(Docker.opened): Check if Docker application is running.
(Docker.open): Start the Docker application.
(Docker.parse_output): Parse list output from various Docker commands.
(Docker.images): List all downloaded images.
(Docker.containers): List all created containers.
(Docker.volumes): List all volumes.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/mocks/subprocess.py:
(Subprocess.completion_generator_for):
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/docker_unittest.py: Added.
(DockerTest.mock_which):
(DockerTest.test_command):
(DockerTest.test_installed):
(DockerTest.test_images):
(DockerTest.test_downloaded):
(DockerTest.test_download):
(DockerTest.test_containers):
(DockerTest.test_created):
(DockerTest.test_create):
(DockerTest.test_running):
(DockerTest.test_start):
(DockerTest.test_stop):
(DockerTest.test_container_decorator):
(DockerTest.test_volume):
(DockerTest.test_volume_create):
(DockerTest.test_volume_remove):
(DockerTest.test_volume_decorator):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9677cac049358081971a5113e5c64de74ab7fbe8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9385 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9661 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9892 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11046 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9257 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9394 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11648 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9623 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12141 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9535 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10450 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8068 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11203 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/9504 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7747 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8559 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15999 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8844 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8709 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12046 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9208 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7497 "Passed tests") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/9325 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8421 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12645 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8970 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0f3a6fc4adabe685ac0a40585b073e9d7cd30e3f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/14753

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e00e9a931815b43209b493f4b0caee7940826dd3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/14753

JonWBedard committed:
[webkitcorepy] Create generic docker utilities
https://bugs.webkit.org/show_bug.cgi?id=257815
rdar://110408458

Reviewed by NOBODY (OOPS!).

Create a generic set of Docker utilities for our various Python services.

* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py:
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/docker.py: Added.
(Docker.Container): Object representing a specific docker container managed from Python.
(Docker.Volume): Object representing a specific docker volume managed from Python.
(Docker.command): Return path to Docker command, if it exists.
(Docker.installed): Check if Docker is installed.
(Docker.opened): Check if Docker application is running.
(Docker.open): Start the Docker application.
(Docker.parse_output): Parse list output from various Docker commands.
(Docker.images): List all downloaded images.
(Docker.containers): List all created containers.
(Docker.volumes): List all volumes.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/mocks/subprocess.py:
(Subprocess.completion_generator_for):
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/docker_unittest.py: Added.
(DockerTest.mock_which):
(DockerTest.test_command):
(DockerTest.test_installed):
(DockerTest.test_images):
(DockerTest.test_downloaded):
(DockerTest.test_download):
(DockerTest.test_containers):
(DockerTest.test_created):
(DockerTest.test_create):
(DockerTest.test_running):
(DockerTest.test_start):
(DockerTest.test_stop):
(DockerTest.test_container_decorator):
(DockerTest.test_volume):
(DockerTest.test_volume_create):
(DockerTest.test_volume_remove):
(DockerTest.test_volume_decorator):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9677cac049358081971a5113e5c64de74ab7fbe8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/14753

dewei-zhu commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/docker.py:
```diff
@@ -0,0 +1,284 @@
+            if port:
+                self.ports[port] = port
+
+        def __enter__(self, cached=True):
```
If `__enter__` is meant for context manager, I'm not aware of a way to run with different `cached` value other than true. 
Also, it looks like almost all functions in this class and `Volume` takes `cached` as argument, I wonder if there is a better way.
